### PR TITLE
Re-introduce io_uring read method

### PR DESCRIPTION
**Pull request by sauliusvl**
### Changelog category:
- Performance Improvement

### Changelog entry:

Add new `local_filesystem_read_method` method `io_uring` based on the asynchronous Linux [io_uring](https://kernel.dk/io_uring.pdf) subsystem, improving read performance almost universally compared to the default `pread` method.

Original PR https://github.com/ClickHouse/ClickHouse/pull/36103 reverted in https://github.com/ClickHouse/ClickHouse/pull/38377, bringing it back for further investigation.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
> Stress test (debug, actions) — Hung check failed
> Stress test (undefined, actions) — Hung check failed

Looks like it is reproducing.

**Comment by alexey-milovidov** [comment]
@sauliusvl let's debug this hanging.

**Comment by sauliusvl** [comment]
added more profile events in order to better understand the issue, since I can't reproduce it locally

**Comment by alexey-milovidov** [comment]
@sauliusvl Let's continue.

**Comment by alexey-milovidov** [comment]
> Hung check failed

It means some queries hung.

**Comment by alexey-milovidov** [comment]
We need this feature :)

**Comment by sauliusvl** [comment]
@alexey-milovidov sorry for my delayed responsiveness, added more error checking/handling and logging, let's see happens.

**Comment by alexey-milovidov** [comment]
@sauliusvl Don't worry, I also respond slowly :)
So, what has happened?

**Comment by alexey-milovidov** [comment]
It is always reproducing. The queries hung. Let's investigate it.

**Comment by alexey-milovidov** [comment]
@sauliusvl Let's not give up.

**Comment by alexey-milovidov** [comment]
> Tests are not finished

This reproduces quite stably.

**Comment by sauliusvl** [comment]
> @sauliusvl Let's not give up.

not giving up yet, just need to find the time to dive deeper, will try to make some time in the near future, once again sorry for the hold up

> This reproduces quite stably

the problem is that it reproduces in CI, but not locally, so it's very time consuming to debug this way - I'll try to find the time and run the full suite locally (if possible)

**Comment by alexey-milovidov** [comment]
@sauliusvl This is a great feature, let's continue :pray:

**Comment by alexey-milovidov** [comment]
This pull request still makes sense :smile:

**Comment by chhetripradeep** [comment]
Just a random thought: can the hang be related with the linux kernel version which we are using in our CI runners (eg. older kernel with some bug in io_uring) while the @sauliusvl local env is running newer kernel. I don't have much io_uring knowledge, but I would like to help. I will also look into it.

**Comment by alexey-milovidov** [comment]
It is possible. We are printing the kernel version at startup in the clickhouse-server logs. Let's check it.

**Comment by chhetripradeep** [comment]
Looking at the [clickhouse server logs](https://s3.amazonaws.com/clickhouse-test-reports/0/774de099cbd4090113344501286dca5463e85d50/stress_test__memory__actions_/clickhouse-server.initial.log) from the hung test  it seems like we are running linux kernel 5.13.0 which seems to be released on [Jun 2021](https://lore.kernel.org/lkml/CAHk-=wj7E9iTGHbqfgtaTAM09WrVzwXjda2_D59MT8D_1=54Rg@mail.gmail.com/).
```
Application: OS name: Linux, version: 5.13.0-1031-aws, architecture: x86_64
```

**Comment by sauliusvl** [comment]
Found a link to this PR while casually reading hacker news comments, now I feel obliged to figure this out :D

**Comment by sauliusvl** [comment]
Correct me if I'm wrong, but three CI runs in a row have now not had a single hung check failure, all current failures seem to be unrelated? 

Full disclaimer: I only rebased on master, so not sure why it suddenly started working 🤷

**Comment by alexey-milovidov** [comment]
Yes, it looks clean. I think it is ready for merge.

We will also be interested in - what will happen if we enable it by default. But it will be more difficult to analyze the test results, so we can do it later.

**Comment by alexey-milovidov** [comment]
> Full disclaimer: I only rebased on master, so not sure why it suddenly started working

It is surprising, and I see two possibilities:
1. Good: We have fixed a generic problem inside the asynchronous reading code, but it was shown mostly with the new IO URing code.
2. Bad: We have accidentally made stress tests less careful.

**Comment by alexey-milovidov** [comment]
I forgot the third possibility: AWS upgraded the Linux kernel version :)

**Comment by chhetripradeep** [comment]
> I forgot the third possibility: AWS upgraded the Linux kernel version :)

I see the previous PR [tests](https://s3.amazonaws.com/clickhouse-test-reports/0/774de099cbd4090113344501286dca5463e85d50/stress_test__memory__actions_/clickhouse-server.initial.log) (which was reverted) were run on kernel v5.13.0 while this PR [tests](https://s3.amazonaws.com/clickhouse-test-reports/38456/ecd2b43cbf4b7280d842eee252f5df3c578fc29c/stress_test__ubsan_/clickhouse-server.initial.log) are running on kernel v5.15.0

One another change is the liburing upstream. Earlier one was from [Jun 20 commit](https://github.com/axboe/liburing/tree/b1735ffa9c5dc5ec18c8fbc121b7b0bc417cdc89) while this one is from [Aug 22 commit](https://github.com/axboe/liburing/tree/316b6de32a7a6e9f569f6e9b0020b6647a614327)

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@sauliusvl one issue related to the changes - Invalid argument while reading from a Log table:
https://s3.amazonaws.com/clickhouse-test-reports/38456/ecd2b43cbf4b7280d842eee252f5df3c578fc29c/stateless_tests__aarch64_.html
Let's check it.

**Comment by sauliusvl** [comment]
> Invalid argument while reading from a Log table

This probably won't be a surprise, but it "works on my machine" :D At first glance the error seems to not be io_uring related, i.e. `cqe->res` is simply the return code of the underlying operation (in this case `read`), which in this case is `-22: "invalid argument"`, so I'm not really sure what could be causing this. It also seems to be happening on `aarch64` only? That's also a bit suspicious.

I've bumped `liburing`, let's see if it persists.

**Comment by chhetripradeep** [comment]
Looking at the powerpc build failure, 
```
Jan 08 21:44:07 /build/contrib/liburing/src/include/liburing/io_uring.h:49:3: error: unknown type name '__kernel_rwf_t'; did you mean '__kernel_off_t'?
Jan 08 21:44:07                 __kernel_rwf_t  rw_flags;
Jan 08 21:44:07                 ^
Jan 08 21:44:07 /build/contrib/sysroot/linux-powerpc64le/powerpc64le-linux-gnu/libc/usr/include/asm-generic/posix_types.h:86:25: note: '__kernel_off_t' declared here
Jan 08 21:44:07 typedef __kernel_long_t __kernel_off_t;
Jan 08 21:44:07                         ^
Jan 08 21:44:07 1 error generated.
```

Do we need to update https://github.com/ClickHouse/sysroot/tree/master/linux-powerpc64le from newer version ?

**Comment by alexey-milovidov** [comment]
One test is broken when we use `io_uring`: https://s3.amazonaws.com/clickhouse-test-reports/38456/98cd2a63ce5759d5086f2ff9433036d884894e8a/stateless_tests__release__s3_storage__[2/2].html

**Comment by alexey-milovidov** [comment]
@chhetripradeep it's easier to disable IO URing on PowerPC because it's not an officially supported platform.

**Comment by sauliusvl** [comment]
> powerpc build failure

I tuned the `liburing` config a bit after upgrading, reverted it

> One test is broken

ugh, once again a very generic error which I can't reproduce locally :disappointed:

**Comment by sauliusvl** [comment]
Looks like all test failures are unrelated this time (two hung checks failed, but seem to be unrelated). Rebasing once again.

**Comment by sauliusvl** [comment]
Only a single unrelated failure this time. Once again, I'm just rebasing on master.

**Comment by sauliusvl** [comment]
OK, so there's quite a lot of `Cannot read from file xxx: errno: 22, strerror: Invalid argument` errors in tests that ran with direct IO, which according to the docs means:

> fd is attached to an object which is unsuitable for reading; or the file was opened with the O_DIRECT flag, and either the address specified in buf, the value specified in count, or the file offset is not suitably aligned.

I'll try to look into it, but this PR re-uses buffer alignment code, so likely something changed there.

**Comment by alexey-milovidov** [comment]
@sauliusvl there are many failed tests:

https://aretestsgreenyet.com/

https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSBjaGVja19uYW1lIExJS0UgJ1N0YXRlJScKICAgIEFORCBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gSU5URVJWQUwgMjQgSE9VUgogICAgQU5EIHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwCiAgICBBTkQgdGVzdF9zdGF0dXMgIT0gJ1NLSVBQRUQnCiAgICBBTkQgdGVzdF9zdGF0dXMgTElLRSAnRiUnCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwpPUkRFUiBCWSBjaGVja19uYW1lLCB0ZXN0X25hbWUsIGNoZWNrX3N0YXJ0X3RpbWU=

https://s3.amazonaws.com/clickhouse-test-reports/0/b512316586c0655b30d6534097df93f3621b57cf/stateless_tests__ubsan_.html

https://s3.amazonaws.com/clickhouse-test-reports/0/e705f11ad1a726deef4d16b9f11e6577c293c9ea/stateless_tests__tsan__[2/3].html

https://s3.amazonaws.com/clickhouse-test-reports/0/283057493b8577e8e16e7f84be6b1f80799fb7bb/stateless_tests__release__databasereplicated__[2/2].html

https://s3.amazonaws.com/clickhouse-test-reports/0/d2e11844d2f84f32ee4e9ec25a0bce7fb95b4b07/stateless_tests__release__databasereplicated__[1/2].html

https://s3.amazonaws.com/clickhouse-test-reports/0/283057493b8577e8e16e7f84be6b1f80799fb7bb/stateless_tests__release__databaseordinary_.html

https://s3.amazonaws.com/clickhouse-test-reports/0/7c6281c446131e66a06ded845a495c963c1e59c2/stateless_tests__release_.html

https://s3.amazonaws.com/clickhouse-test-reports/0/8dd6dbc4527d490fb1615d3f44d218c000a4a247/stateless_tests__aarch64_.html

**Comment by alexey-milovidov** [comment]
All of them in the form:

```
Failed reading 4096 bytes at offset 0 ... errno: 22, strerror: Invalid argument: ... While executing Log.
```

**Comment by alexey-milovidov** [comment]
But at least one of them is:

```
Failed reading 4096 bytes at offset 0 ... errno: 22, strerror: Invalid argument: ... While executing StripeLog.
```

**Comment by alexey-milovidov** [comment]
Failed tests are:

01502_long_log_tinylog_deadlock_race
01632_tinylog_read_write

Both are doing SELECT and INSERT with high concurrency.

**Comment by alexey-milovidov** [comment]
Worth noting the fact that Log, TinyLog, and StripeLog are writing with appending to files (unlike MergeTree, which is immutable). Also, the `File` storage is append-able, as well as the backup data of `Set` and `Join`.

**Comment by alexey-milovidov** [comment]
What happens, if you append to a file of size smaller than 4096 using normal writes and in another thread, read a block of up to 4096 bytes?

**Comment by sauliusvl** [comment]
I'll try to reproduce it. In theory (and to my understanding) the semantics of `io_uring` should be identical to those of the underlying operations, in this case `read`, just that they're asynchronous, so maybe there is some race happening.

File: src/Disks/IO/createReadBufferFromFileBase.cpp [hunk 1]
```diff
@@ -80,6 +82,19 @@ std::unique_ptr<ReadBufferFromFileBase> createReadBufferFromFileBase(
         {
             res = std::make_unique<ReadBufferFromFilePReadWithDescriptorsCache>(filename, buffer_size, actual_flags, existing_memory, buffer_alignment, file_size);
         }
+        else if (settings.local_fs_method == LocalFSReadMethod::io_uring)
+        {
+#if defined(OS_LINUX)
+            static std::shared_ptr<IOUringReader> reader = std::make_shared<IOUringReader>(512);
```
Comment (azat) [inline src/Disks/IO/createReadBufferFromFileBase.cpp:88]: It is not possible to join the thread with static variable, this should be moved to global context, like other things (like thread pool for reading and so on).

**Comment by binmahone** [comment]
hello @sauliusvl  @alexey-milovidov 

I'm testing local_filesystem_read_method=io_uring on my local workstation. And surprisingly find that io_uring is 2 times slower than pread_threadpool or mmap. 

**This is somewhat contradictory to the conclusion stated in the https://github.com/ClickHouse/ClickHouse/pull/36103**. Have you received similar feedback?  Or am I testing it in a wrong way?

```
ws2 :) SET local_filesystem_read_method = 'mmap';

SET local_filesystem_read_method = 'mmap'

Query id: 183d70e9-ff4f-4d42-a5be-642b0103c9a6

Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 23.11.3.

Ok.

0 rows in set. Elapsed: 0.013 sec. 

ws2 :) select min(ignore(*)) from lineitem5;

SELECT min(ignore(*))
FROM lineitem5

Query id: 47c162af-e504-4e52-8dd3-cedccf0676f7

┌─min(ignore(l_orderkey, l_partkey, l_suppkey, l_linenumber, l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate, l_commitdate, l_receiptdate, l_shipinstruct, l_shipmode, l_comment))─┐
│                                                                                                                                                                                                                       0 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 5.347 sec. Processed 600.04 million rows, 99.32 GB (112.21 million rows/s., 18.57 GB/s.)
Peak memory usage: 97.81 MiB.

ws2 :) SET local_filesystem_read_method = 'pread_threadpool';

SET local_filesystem_read_method = 'pread_threadpool'

Query id: 8708c8b4-ad20-4610-ad5f-e318bbf06606

Ok.

0 rows in set. Elapsed: 0.001 sec. 

ws2 :) select min(ignore(*)) from lineitem5;

SELECT min(ignore(*))
FROM lineitem5

Query id: 1631826a-0979-4366-aa8b-d56fe5cf9278

┌─min(ignore(l_orderkey, l_partkey, l_suppkey, l_linenumber, l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate, l_commitdate, l_receiptdate, l_shipinstruct, l_shipmode, l_comment))─┐
│                                                                                                                                                                                                                       0 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 5.111 sec. Processed 600.04 million rows, 99.32 GB (117.41 million rows/s., 19.43 GB/s.)
Peak memory usage: 97.83 MiB.

ws2 :) SET local_filesystem_read_method = 'io_uring';

SET local_filesystem_read_method = 'io_uring'

Query id: 400714f3-fe3d-488e-baa7-c3aebe3718b7

Ok.

0 rows in set. Elapsed: 0.000 sec. 

ws2 :) select min(ignore(*)) from lineitem5;

SELECT min(ignore(*))
FROM lineitem5

Query id: 3ae698c4-5e2c-4af6-b476-ec9576c626a0

┌─min(ignore(l_orderkey, l_partkey, l_suppkey, l_linenumber, l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate, l_commitdate, l_receiptdate, l_shipinstruct, l_shipmode, l_comment))─┐
│                                                                                                                                                                                                                       0 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 11.136 sec. Processed 600.04 million rows, 99.32 GB (53.88 million rows/s., 8.92 GB/s.)
Peak memory usage: 126.12 MiB.

```


More details about test data and hardware:

1. Lineitem5 is a mergetree table is from TPCH100, with mergetree size around 30G
2. CPU: 13th Gen Intel(R) Core(TM) i9-13900K 
3. RAM: 128G RAM
4. Disk: Samsung SSD 980 1TB

**Comment by alexey-milovidov** [comment]
I didn't observe io_uring be much slower, but also I have no big expectations because I wasn't able to find cases when it is faster.

**Comment by binmahone** [comment]
> I didn't observe io_uring be much slower, but also I have no big expectations because I wasn't able to find cases when it is faster.

Yeah. In my test the TPCH table can fit into page cache. So I expect a comparable result between io_uring and others. This is wield.

I guess io_uring is better suited for cases where random read requests dominate. In CH, for each working read thread, it's basically doing sequential read with double buffering, this cannot accumulate enough parallel IO requests. Please correct me if I'm wrong.

**Comment by sauliusvl** [comment]
When I was working on this originally 1.5 years ago I observed ~10% performance improvements for HDD workloads when data was not in the page cache, as shown in the original PR. When data was in the page cache OR data was on SSD it was similar or sometimes slower up to ~3%, but I never observed such drastic differences. When I re-opened this PR ~1 year ago even the HDD/no-cache gains were essentially not visible anymore - maybe the old read methods got optimized even further in that time?

Anyway, my original motivation was speeding up parallel non-cached HDD workloads, as I expected `io_uring` to do a better job of ordering requests in the background. For SDD/cached workloads I believe the machinery of managing SQEs in std::maps, checking for overflows, queuing and things like that start becoming a bottleneck and the whole read method becomes slower compared to others. In order to gain the benefits of `io_uring` for these workloads it should be implemented a bit differently, e.g. it could use polling mode, submit/consume SQEs in bigger batches/single syscall, etc. It's in my TODO list to try optimizing it in this direction.

**Comment by binmahone** [comment]
> When I was working on this originally 1.5 years ago I observed ~10% performance improvements for HDD workloads when data was not in the page cache, as shown in the original PR. When data was in the page cache OR data was on SSD it was similar or sometimes slower up to ~3%, but I never observed such drastic differences. When I re-opened this PR ~1 year ago even the HDD/no-cache gains were essentially not visible anymore - maybe the old read methods got optimized even further in that time?
> 
> Anyway, my original motivation was speeding up parallel non-cached HDD workloads, as I expected `io_uring` to do a better job of ordering requests in the background. For SDD/cached workloads I believe the machinery of managing SQEs in std::maps, checking for overflows, queuing and things like that start becoming a bottleneck and the whole read method becomes slower compared to others. In order to gain the benefits of `io_uring` for these workloads it should be implemented a bit differently, e.g. it could use polling mode, submit/consume SQEs in bigger batches/single syscall, etc. It's in my TODO list to try optimizing it in this direction.

thanks for the explanation ! Should you have any updates on this direction, can you please kindly AT me, thanks!

**Pull request merged**